fastly/data_source_fastly_tls_activation.go

```diff
@@ -0,0 +1,167 @@
+	return &schema.Resource{
+		Read: dataSourceFastlyTLSActivationRead,
+		Schema: map[string]*schema.Schema{
+			"id": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Computed:      true,
+				Description:   "Fastly Activation ID. Conflicts with all other filters.",
+				ConflictsWith: []string{"certificate_id", "configuration_id", "domain"},
+			},
+			"certificate_id": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Computed:      true,
+				ConflictsWith: []string{"id"},
+				Description:   "ID of the TLS Certificate used.",
+			},
+			"configuration_id": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Computed:      true,
+				ConflictsWith: []string{"id"},
+				Description:   "ID of the TLS Configuration used.",
+			},
+			"domain": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Computed:      true,
+				ConflictsWith: []string{"id"},
+				Description:   "Domain that TLS was enabled on.",
+			},
+			"created_at": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Timestamp (GMT) when TLS was enabled.",
+			},
+		},
+	}
+}
+
+func dataSourceFastlyTLSActivationRead(d *schema.ResourceData, meta interface{}) error {
+
+	conn := meta.(*FastlyClient).conn
+
+	var activation *fastly.TLSActivation
+
+	if v, ok := d.GetOk("id"); ok {
+
+		foundActivation, err := conn.GetTLSActivation(&fastly.GetTLSActivationInput{
+			ID: v.(string),
+		})
+		if err != nil {
+			return err
+		}
+		activation = foundActivation
+
+	} else {
+
+		filters := getTLSActivationFilters(d)
+
+		activations, err := listTLSActivations(conn, filters...)
+		if err != nil {
+			return err
+		}
+
+		if len(activations) == 0 {
+			return fmt.Errorf("Your query returned no results. Please change your search criteria and try again")
+		}
+
+		if len(activations) > 1 {
+			return fmt.Errorf("Your query returned more than one result. Please change to a more specific search criteria")
+		}
+
+		activation = activations[0]
+	}
+
+	return dataSourceFastlyTLSActivationSetAttributes(activation, d)
+}
+
+type TLSActivationPredicate func(activation *fastly.TLSActivation) bool
+
+func getTLSActivationFilters(d *schema.ResourceData) []TLSActivationPredicate {
+	var filters []TLSActivationPredicate
+
+	if v, ok := d.GetOk("certificate_id"); ok {
+		filters = append(filters, func(c *fastly.TLSActivation) bool {
+			return c.Certificate.ID == v.(string)
+		})
+	}
+	if v, ok := d.GetOk("configuration_id"); ok {
+		filters = append(filters, func(c *fastly.TLSActivation) bool {
+			return c.Configuration.ID == v.(string)
+		})
+	}
+	if v, ok := d.GetOk("domain"); ok {
+		filters = append(filters, func(c *fastly.TLSActivation) bool {
+			return c.Domain.ID == v.(string)
+		})
+	}
+
+	return filters
+}
+
+func listTLSActivations(conn *fastly.Client, filters ...TLSActivationPredicate) ([]*fastly.TLSActivation, error) {
+	var activations []*fastly.TLSActivation
```

Comment:
This is a good point, yes. In the case of TLS activations we do not share the `list*` function between the singular and plural data sources (because the plural one uses an API filter directly whereas the others do all the filtering after pulling the results down). Therefore in this case, we could just modify this function to break the loop early once the filter has passed. However, in most (if not all) of the other data sources, we do share the `list*` function between singular and plural data sources, to abstract away the API's inconsistent pagination methods, so we might lose the consistency among the data sources in this case.

Do you think the consistency with other data sources takes precedence, or should we take the easy code optimisation here while we can?

Reply:
I think consistency is fine in this case. If there's a performance issue that arises we can address it when it happens.